src/message.rs

```diff
@@ -93,34 +99,38 @@ pub trait Message: Debug + Send + Sync {
         Self: Default,
     {
         let mut message = Self::default();
-        message.merge_length_delimited(buf)?;
+        message.merge_length_delimited(buf, &mut DecodeContext::default())?;
         Ok(message)
     }
 
     /// Decodes an instance of the message from a buffer, and merges it into `self`.
     ///
     /// The entire buffer will be consumed.
-    fn merge<B>(&mut self, buf: B) -> Result<(), DecodeError>
+    fn merge<B>(&mut self, buf: B, ctx: &mut DecodeContext) -> Result<(), DecodeError>
```

Comment:
This is part of the public API, so it shouldn't have a context param, it should be handled the same as you did `decode_length_delimited` above.

Reply:
In other words, `merge_field` should be the only method with a context param.

Reply:
That doesn't work unfortunately because merge is called recursively so if we don't include the context, then we lose track of the recursion count